pkg/ipcache/gc.go

```diff
@@ -41,15 +43,22 @@ func newAsyncPrefixReleaser(parent prefixReleaser, interval time.Duration) *asyn
 		Name:        "ipcache-identity-gc",
 		MinInterval: interval,
 		TriggerFunc: func(reasons []string) {
-			// TODO: Structure the code to pass context down
-			//       from the Daemon.
-			ctx, cancel := context.WithTimeout(
-				context.TODO(),
+			innerCtx, cancel := context.WithTimeout(
```

Comment:
I'm a bit wary of carrying multiple contexts. Although the inner context is derived from the daemon context, see the comment below. If that works we need not have multiple contexts.

Reply:
In theory all contexts should derive from the top of the context tree so that you can cancel the top one and cancel all logic where that context or subcontexts are used. The way it didn't derive from the daemon before was weird, since it meant that the child context could outlive the overall agent. I can understand some reluctance around 'multiple contexts' if some code is trying to force two different parent contexts, since this isn't really handled well in the standard library. However, this is a pretty standard way to derive a context from a parent context. Maybe the only unusual thing here is that usually in the codebase we'd call one 'parentCtx' and one 'ctx'.

Reply:
^ strictly speaking we could avoid doing the actual shutdown with this context, and instead do an explicit shutdown.. but even with that I think that this context should be derived from the agent context. So presume there's some logic going on using this context, first the agent context gets cancelled and we force this logic to cancel & return to caller. This also means that all subsequent calls with the ctx will fail quickly. Then separately the agent calls the explicit Shutdown() to ensure the goroutines stop etc. and nothing remains running.